tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php

```diff
@@ -49,10 +73,71 @@ public function __construct(
 
 	/**
 	 * @Given /^the user has activated TOTP Second\-factor auth but not verified$/
+	 * @When /^the user activates TOTP Second\-factor auth but does not verify$/
 	 *
 	 * @return void
 	 */
 	public function theUserHasActivatedTOTPSecondFactorAuthButNotVerified() {
 		$this->personalSecuritySettingsPage->activateTOTP();
+		$this->totpSecret = $this->personalSecuritySettingsPage->getSecretCode();
+	}
+
+	/**
+	 * Returns secret code extracted from QRCode
+	 *
+	 * @return string
+	 */
+	public function getSecretCodeFromQRCode() {
+		$path = \tempnam(\sys_get_temp_dir(), 'totp_qrcode');
+		$data = \explode(',', $this->personalSecuritySettingsPage->getQRCode());
+		$file = \fopen($path, 'wb');
+		\fwrite($file, \base64_decode($data[1]));
+		$qrCode = new QrReader($path);
+		\preg_match('/secret=(?P<secret>[A-Z0-9]{16})/', $qrCode->text(), $matches);
+		return $matches['secret'];
+	}
+
+	/**
+	 * @When the user adds one-time key generated from the secret key using the webUI
+	 *
+	 * @return void
+	 */
+	public function theUserAddsVerficationKeyFromSecretKeyUsingWebUI() {
+		$remainder = \time() % 30;
+		if ($remainder % 30 < 10) {
```

Comment:
``remainder`` is already ``% 30``` - doing it again is not needed.

Reply:
Oops, this has to be:
```suggestion
		if ($remainder < 10) {
```

As per @karakayasemi, we don't need to sleep anyway. But, we do need to have some state to know that we have already used the key. How should we approach that? Sleeping for 30 sec would be very bad.